changelogs/unreleased/popen-mac-os-race.md

```diff
@@ -0,0 +1,12 @@
+## bugfix/lua/popen
+
+* Handle possible race condition between `setgprp()` and `killpg()` on Mac OS
+  12 and newer (gh-6674).
```

Comment:
This shouldn't close #6674 - the ticket is about getting rid of `vfork` usage, which this PR clearly doesn't do. We shouldn't use `vfork`, because it'll probably be dropped in OS X soon.

Overall, I don't like this patch, because it looks like a hack. I think ~~we should drop OS X support instead~~ we can ignore this bug, because nobody complained, and instead come up with a proper fix for OS X. Technically, the patch looks correct though, but I wouldn't commit it, since there's no pressure to fix this bug.

Reply:
I don't like non-working functionality on one of supported OS. My goal is to make popen working on Mac OS. I agree that my way to solve the problem is hacky, but it is the best I can propose now (without massive redesigning/refactoring at least).

(I'm thinking on possible massive rework -- using forkserver, using exec wrapper and so on -- but it is not what I ready to deliver in a tight time. If we'll look into `posix_spawn()` direction, there are stones on this way: say, `posix_spawn()` will not give as full control of what to do in the child and, sadly, it is not specified by POSIX, whether atfork handlers are called. Unlike `vfork()`, which is deprecated and may be deleted, bla-bla, but works now. Usual story on Mac OS: see for example https://metacpan.org/dist/EV/view/libev/ev.pod#OS/X-AND-DARWIN-BUGS.)

We can discuss your understanding of a proper fix and estimate cons and pros. But I can't agree with 'let it be broken' position, when there are ways to heal it.

I'm closing #6674, because getting rid of `vfork()` is not the real goal and it is unclear, whether we should do so. I can open another issue to fix popen on Mac OS and close it from the commit. I would also close #6674, because it has no much sense if everything works.

Reply:
We discussed it with Vladimir and Igor M. and agreed on the following points:

* Mac OS is the problematic platform, but the problem should be fixed.
* Leave #6674 open for possible future redesign, file a separate issue regarding this particular Mac OS problem (filed #7658) and close it from this PR.
* Decouple this patch it from #7648: don't use the fiber cond.
* Move the group leadership waiting to `popen.new()` instead of `:send_signal()`. It is unexpected that `:send_signal()` performs some waiting and it is undesirable to extend its contract with error situations that may appear in this new code.
* Don't add yields into the `popen.new()` implementation, it may break user's code. Perform a real sleep in the parent process instead. At least it'll not be much worse than `vfork()` on Linux, which freezes the parent process until `execve()`.

We discussed the following ways to implement child's `execve()` waiting:

* Using `pipe()`. It would be nice, but costs a file descriptor. It is still unclear, whether it worth it. Decided to don't proceed this way, at least within the scope of the bug.
* Using a signal from the child to the parent. It would be nice too, but it is really unclear, which signal we can consume to don't break any existing or future code (including users' code). We should take care to the fact that a user may load its own dynamic libraries. Doubtful way and I'll not implement it in the scope of the bugfix.

It seems, it worth to implement a simple polling at least in the scope of the bugfix. I'll do.

Reply:
Update the patch according to our agreements. Tested on Linux and Mac OS, seems to work as expected.